### DISCUSSION TITLE
redirect plugin configuration if I rename url

### DISCUSSION THREAD

**Discussion by luweizheng** [vuepress-theme-hope/vuepress-theme-hope#3100] [Q&A] [answered]
Hi all,

I rename a folder from `machine-learning` to `deep-learning`. I want to configure redirect plugin so that old URLs can be redirected to the new URLs and old URLs have already been indexed by search engines.

Here is the folder structure where `deep-learning` is formerly `machine-learning` before.

```
└── src
    ├── README.md
    ├── about.md
    ├── blog
    ├── deep-learning <- formerly: machine-learning
    ├── donate.md
    ├── flink
    ├── gpu
    └── python
```

Here is the redirect plugin setting of `config.ts`:

```
import { redirectPlugin } from "vuepress-plugin-redirect";

plugins: [
    [
      // ...
      redirectPlugin({
        config: (app) =>
          Object.fromEntries(
            app.pages
              .filter(({ path }) => path.startsWith("/deep-learning/"))
              .map(({ path }) => [path.replace(/^\/deep-learning\//, "/machine-learning/"), path])
          ),
      }),
    ],
  ],
```

But it does not work? What's the problem of my configuration? Any suggestions?

**Comment by Mister-Hope** [comment]
Reproduction site and repo is required for these kind of things, unless the demo itself is not working.

**Reply by luweizheng** [reply]
Just push my recent changes into https://github.com/luweizheng/lulaoshi.info
But I get the following error log when building:

```
✔ Rendering 124 pages - done in 2.60s
vuepress-plugin-seo2:  ✔ Generating robots.txt
vuepress-plugin-sitemap2:  ✔ Generating sitemap to /sitemap.xml
vuepress-plugin-sitemap2:  ✔ Appended sitemap path to robots.txt
✖ Generating redirect files - failed in 3ms
error error in hook onGenerated from vuepress-plugin-redirect
Error: EISDIR: illegal operation on a directory, open '/Users/luweizheng/Projects/web/lulaoshi.info/src/.vuepress/dist/machine-learning/attention'
```

**Reply by luweizheng** [reply]
This error message also appears in my CI/CD workflow.

**Reply by Mister-Hope** [reply, marked as answer]
~~You should use `/a/b/index.html` instead of `/a/b/`.~~

~~Docs are updated at https://github.com/vuepress-theme-hope/vuepress-theme-hope/commit/a59a590cb3db7be1ebd0380ca191b791c0d14464~~

We made some improvements at https://github.com/vuepress-theme-hope/vuepress-theme-hope/commit/6fbea5062bb7fadda00a78856afb833bbe27e037

**Reply by luweizheng** [reply]
Cool. But your improvements need some document updates.

Another question:

I move an article from `/posts` which is a blog page to `/python` which is a document page.

Here is my markdown: https://github.com/luweizheng/lulaoshi.info/blob/main/src/python/setup-py/entry-points.md

Now the redirect plugin cannot redirect from blog posts to documents.

**Reply by luweizheng** [reply]
@Mister-Hope Hi, could you check why I cannot redirect from blog posts to documents?

**Reply by Mister-Hope** [reply]
That should work, can you try to provide a minimal reproduction?

**Reply by luweizheng** [reply]
> https://github.com/luweizheng/lulaoshi.info/blob/main/src/python/setup-py/entry-points.md

I put `redirectFrom` my `frontmatter` in this document page.

**Reply by Mister-Hope** [reply]
view-source:https://lulaoshi.info/blog/python-setup-entry-points/index.html

WTF is this? (Doesn't mean to you, just the file) Why is it different from my local result? Are you sure you are using the correct deployment action?

**Reply by Mister-Hope** [reply]
This is my local result:

```html
<!DOCTYPE html>
<html lang="zh">
<head>
  <meta charset="UTF-8">
  <meta name="robots" content="noindex">
  <meta http-equiv="refresh" content="0; url=/python/setup-py/entry-points.html">
  <link rel="canonical" href="/python/setup-py/entry-points.html">
  <title>Redirecting...</title>
  <script>
    const anchor = window.location.hash.substring(1);
    location.href = `/python/setup-py/entry-points.html${anchor? `#${anchor}`: ""}`;
  </script>
</head>
<body>
  <p>Redirecting...</p>
</body>
</html>
```

So tell me why is it different from your site? Next time, a minimal reproduction is required when I ask that, otherwise further contact will all be ignored.

**Reply by luweizheng** [reply]
How to get the local redirect file like what you paste? 

view-source:http://localhost:8080/blog/python-setup-entry-points/index

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script type="module" src="[/@vite/client](view-source:http://localhost:8080/@vite/client)"></script>

    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
  </head>
  <body>
    <div id="app"></div>
  <script type="module" src="[/index.html?html-proxy&index=0.js](view-source:http://localhost:8080/index.html?html-proxy&index=0.js)"></script>
</body>
</html>
```

In my local development environment, http://localhost:8080/blog/python-setup-entry-points/index can redirect. Why there is an `index` at the end of the URL? Why http://localhost:8080/blog/python-setup-entry-points/ can not work?

And I :
1. change package manager from `npm` to `pnpm`
2. change the my github action workflow as the workflow in your personal blog site
3. remove all the files in the server and re-rsync the `dist` folder to the server

Still get the same result.

**Reply by Mister-Hope** [reply]
The plugin behaves differently in dev and build mode, and the reply content is what I get from my local build result. Try to get started with a demo from https://stackblitz.com/fork/vuepress-theme-hope, and add the plugin and file to see if it works correctly, if so check your config.

Or more clearly, this is the result if javascript is disabled:

<img width="756" alt="image" src="https://user-images.githubusercontent.com/33315834/236637565-1ca6061d-f085-4310-bb68-7a296ab366b5.png">

And this is not what I am getting when building localy

**Comment by Mister-Hope** [comment]
> Why there is an index at the end of the URL?

Your server perform redirects from https://lulaoshi.info/blog/python-setup-entry-points/index.html to https://lulaoshi.info/blog/python-setup-entry-points/index, and that has nothing to do with me.

**Comment by luweizheng** [comment]
Maybe the apache or other configuration.

But I give it up ...
